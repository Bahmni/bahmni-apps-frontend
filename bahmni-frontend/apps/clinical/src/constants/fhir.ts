export const FHIR_CONCEPT_CLASS_EXTENSION_URL =
  'http://fhir.bahmni.org/ext/ValueSet/concept-class';
