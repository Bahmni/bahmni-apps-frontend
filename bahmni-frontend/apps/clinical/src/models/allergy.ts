import type { Coding } from 'fhir/r4';

/**
 * Interface representing an allergy input entry for form handling
 */
export interface AllergyInputEntry {
  id: string;
  display: string;
  type: string;
  selectedSeverity: Coding | null;
  selectedReactions: Coding[];
  note?: string;
  errors: {
    severity?: string;
    reactions?: string;
  };
  hasBeenValidated: boolean;
}

export type AllergenType = 'food' | 'medication' | 'environment';

export interface AllergenConcept {
  uuid: string;
  display: string;
  type: AllergenType | null;
  disabled?: boolean;
}
