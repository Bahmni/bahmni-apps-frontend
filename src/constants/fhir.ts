export const FHIR_ENCOUNTER_TYPE_CODE_SYSTEM =
  'http://fhir.openmrs.org/code-system/encounter-type';
export const HL7_CONDITION_CATEGORY_CONDITION_CODE = 'problem-list-item';
