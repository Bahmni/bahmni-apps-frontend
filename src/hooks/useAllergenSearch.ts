import { useEffect, useState, useMemo } from 'react';
import useDebounce from './useDebounce';
import { fetchAndFormatAllergenConcepts } from '@services/allergenService';
import { AllergenConcept } from '@types/concepts';
import { getFormattedError } from '@utils/common';

interface UseAllergenSearchResult {
  allergens: AllergenConcept[];
  isLoading: boolean;
  error: Error | null;
}

/**
 * A hook that provides debounced search functionality for allergen concepts.
 * It eagerly loads all allergen concepts and filters them based on the search term.
 * @param searchTerm - Optional search term to filter allergens
 * @returns Object containing filtered allergens, loading state, and any errors
 */
const useAllergenSearch = (
  searchTerm: string = '',
): UseAllergenSearchResult => {
  const [allergens, setAllergens] = useState<AllergenConcept[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const debouncedSearchTerm = useDebounce(searchTerm);

  // Load all allergens on mount
  useEffect(() => {
    const loadAllergens = async () => {
      setIsLoading(true);
      setError(null);
      try {
        const concepts = await fetchAndFormatAllergenConcepts();
        setAllergens(concepts);
      } catch (err) {
        const formattedError = getFormattedError(err);
        setError(new Error(formattedError.message));
      } finally {
        setIsLoading(false);
      }
    };
    loadAllergens();
  }, []);

  // Filter allergens based on search term
  const filteredAllergens = useMemo(() => {
    if (!allergens.length) return [];

    const searchTermLower = debouncedSearchTerm?.toLowerCase().trim() || '';
    if (!searchTermLower) return allergens;

    return allergens.filter((allergen) => {
      const displayLower = allergen.display.trim().toLowerCase();
      return allergen && displayLower.includes(searchTermLower);
    });
  }, [allergens, debouncedSearchTerm]);

  return {
    allergens: filteredAllergens,
    isLoading,
    error,
  };
};

export default useAllergenSearch;
