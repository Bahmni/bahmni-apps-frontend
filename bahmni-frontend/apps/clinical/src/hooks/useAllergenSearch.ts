import { Coding } from 'fhir/r4';
import { useEffect, useState, useMemo } from 'react';
import {
  fetchAndFormatAllergenConcepts,
  fetchReactionConcepts,
  getFormattedError,
  useTranslation,
} from '@bahmni-frontend/bahmni-services';
import { AllergenConcept } from '../models/allergy';
import { useClinicalConfig } from './useClinicalConfig';
import useDebounce from './useDebounce';

interface UseAllergenSearchResult {
  allergens: AllergenConcept[];
  reactions: Coding[];
  isLoading: boolean;
  error: Error | null;
}

/**
 * A hook that provides debounced search functionality for allergen concepts.
 * It eagerly loads all allergen concepts and filters them based on the search term.
 *
 * @param searchTerm - Optional search term to filter allergens
 * @returns Object containing filtered allergens, loading state, and any errors
 */
const useAllergenSearch = (serchTerm: string = ''): UseAllergenSearchResult => {
  const [allergens, setAllergens] = useState<AllergenConcept[]>([]);
  const [reactions, setReactions] = useState<Coding[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const debouncedSearchTerm = useDebounce(serchTerm);

  const {
    clinicalConfig,
    isLoading: isConfigLoading,
    error: configError,
  } = useClinicalConfig();
  const { t } = useTranslation();

  // Load all allergens when config is available
  useEffect(() => {
    const loadAllergens = async () => {
      setIsLoading(true);
      setError(null);

      if (configError) {
        setError(configError);
        setIsLoading(false);
        return;
      }

      // Wait for config to load
      if (isConfigLoading) {
        return;
      }

      // Check if config is available
      if (!clinicalConfig) {
        setError(new Error(t('ERROR_CLINICAL_CONFIG_NOT_FOUND')));
        setIsLoading(false);
        return;
      }

      // Check if consultationPad is available
      if (!clinicalConfig.consultationPad) {
        setError(new Error(t('ERROR_CONSULTATION_PAD_NOT_FOUND')));
        setIsLoading(false);
        return;
      }

      // Check if allergyConceptMap is available
      if (!clinicalConfig.consultationPad.allergyConceptMap) {
        setError(new Error(t('ERROR_ALLERGY_CONCEPT_MAP_NOT_FOUND')));
        setIsLoading(false);
        return;
      }

      const {
        medicationAllergenUuid,
        foodAllergenUuid,
        environmentalAllergenUuid,
        allergyReactionUuid,
      } = clinicalConfig.consultationPad.allergyConceptMap;

      try {
        const allergens = await fetchAndFormatAllergenConcepts(
          medicationAllergenUuid,
          foodAllergenUuid,
          environmentalAllergenUuid,
        );
        setAllergens(allergens);
        const reactions = await fetchReactionConcepts(allergyReactionUuid);
        setReactions(reactions);
      } catch (err) {
        const formattedError = getFormattedError(err);
        setError(new Error(formattedError.message));
      } finally {
        setIsLoading(false);
      }
    };
    loadAllergens();
  }, [clinicalConfig, isConfigLoading, configError]);

  // Filter allergens based on search term
  const filteredAllergens = useMemo(() => {
    if (!allergens.length) return [];

    const searchTermLower = debouncedSearchTerm?.toLowerCase().trim() || '';
    if (!searchTermLower) return allergens;

    // Split search term into words for more flexible matching
    const searchWords = searchTermLower.split(/\s+/);

    return allergens.filter((allergen) => {
      const displayLower = allergen.display.toLowerCase();
      // Match if any search word is found anywhere in the display name
      return searchWords.some((word) => displayLower.includes(word));
    });
  }, [allergens, debouncedSearchTerm]);

  return {
    allergens: filteredAllergens,
    reactions,
    isLoading,
    error,
  };
};

export default useAllergenSearch;
