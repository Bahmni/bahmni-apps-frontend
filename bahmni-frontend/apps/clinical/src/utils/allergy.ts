import { AllergenType } from '../models/allergy';

/**
 * Maps allergy types to their corresponding i18n translation keys
 * @param type - The type of allergen
 * @returns The i18n translation key for the allergen type
 */
export const getCategoryDisplayName = (
  type: AllergenType | string | undefined | null,
): string => {
  if (type === undefined) return 'undefined';
  if (type === null) return 'null';

  const typeToI18nKey: Record<AllergenType, string> = {
    food: 'ALLERGY_TYPE_FOOD',
    medication: 'ALLERGY_TYPE_DRUG',
    environment: 'ALLERGY_TYPE_ENVIRONMENT',
  };
  return typeToI18nKey[type as AllergenType] || type;
};
