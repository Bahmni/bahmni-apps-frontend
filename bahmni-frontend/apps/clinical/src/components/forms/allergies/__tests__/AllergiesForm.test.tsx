import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Coding } from 'fhir/r4';
import { axe, toHaveNoViolations } from 'jest-axe';
import useAllergenSearch from '../../../../hooks/useAllergenSearch';
import { useAllergyStore } from '../../../../stores/allergyStore';
import { AllergenConcept } from '../../../../types/allergy';
import AllergiesForm from '../AllergiesForm';

expect.extend(toHaveNoViolations);

// Mock modules
jest.mock('../../../../stores/allergyStore');
jest.mock('../../../../hooks/useAllergenSearch');
jest.mock('../styles/AllergiesForm.module.scss', () => ({
  allergiesFormTile: 'allergiesFormTile',
  allergiesFormTitle: 'allergiesFormTitle',
  allergiesBox: 'allergiesBox',
  selectedAllergyItem: 'selectedAllergyItem',
}));

const mockAllergen: AllergenConcept = {
  uuid: 'test-allergy-1',
  display: 'Peanut Allergy',
  type: 'food',
  disabled: false,
};

const mockReactions: Coding[] = [
  {
    code: 'hives',
    display: 'REACTION_HIVES',
    system: 'http://snomed.info/sct',
  },
  {
    code: 'rash',
    display: 'REACTION_RASH',
    system: 'http://snomed.info/sct',
  },
];

const mockSelectedAllergy = {
  id: mockAllergen.uuid,
  display: mockAllergen.display,
  selectedSeverity: null,
  selectedReactions: [],
  errors: {},
  hasBeenValidated: false,
};

const mockAllergyStore = {
  selectedAllergies: [],
  addAllergy: jest.fn(),
  removeAllergy: jest.fn(),
  updateSeverity: jest.fn(),
  updateReactions: jest.fn(),
  updateNote: jest.fn(),
  validateAllAllergies: jest.fn(),
  reset: jest.fn(),
  getState: jest.fn(),
};

const mockAllergenSearch = {
  allergens: [],
  reactions: mockReactions,
  isLoading: false,
  error: null,
};

// Test utilities
const renderAllergiesForm = (overrides = {}) => {
  const mockStore = { ...mockAllergyStore, ...overrides };
  (
    useAllergyStore as jest.MockedFunction<typeof useAllergyStore>
  ).mockReturnValue(mockStore);

  return render(
      <AllergiesForm />
  );
};

const mockAllergenSearchHook = (overrides = {}) => {
  const searchHook = { ...mockAllergenSearch, ...overrides };
  (
    useAllergenSearch as jest.MockedFunction<typeof useAllergenSearch>
  ).mockReturnValue(searchHook);
};

const getSearchCombobox = () =>
  screen.getByRole('combobox', { name: /search for allergies/i });

describe('AllergiesForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(console, 'error').mockImplementation(() => {});
    window.HTMLElement.prototype.scrollIntoView = jest.fn();

    // Set default mocks
    (
      useAllergyStore as jest.MockedFunction<typeof useAllergyStore>
    ).mockReturnValue(mockAllergyStore);
    mockAllergenSearchHook();
  });

  describe('Rendering', () => {
    it('should render search box correctly', () => {
      renderAllergiesForm();
      expect(getSearchCombobox()).toBeInTheDocument();
    });

    it('should display selected allergies', () => {
      renderAllergiesForm({ selectedAllergies: [mockSelectedAllergy] });
      expect(screen.getByText(/Peanut Allergy/)).toBeInTheDocument();
    });

    it('should show loading state while searching', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ isLoading: true });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'a');

      await waitFor(() => {
        expect(screen.getByText(/loading concepts/i)).toBeInTheDocument();
      });
    });

    it('should show error when search fails', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ error: new Error('Failed to load allergens') });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'a');

      await waitFor(() => {
        expect(
          screen.getByText(/unexpected error occurred/i),
        ).toBeInTheDocument();
      });
    });

    it('should show message when no search results found', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'nonexistent');

      await waitFor(() => {
        expect(
          screen.getByText(/No matching allergen recorded/i),
        ).toBeInTheDocument();
      });
    });
  });

  describe('Allergy Selection', () => {
    it('should add an allergy when selected from search results', async () => {
      const user = userEvent.setup();
      const mockAddAllergy = jest.fn();

      mockAllergenSearchHook({ allergens: [mockAllergen] });
      renderAllergiesForm({ addAllergy: mockAddAllergy });

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(() => {
        expect(screen.getByText('Peanut Allergy [Food]')).toBeInTheDocument();
      });

      await user.click(screen.getByText('Peanut Allergy [Food]'));

      await waitFor(() => {
        expect(mockAddAllergy).toHaveBeenCalledWith(mockAllergen);
      });
    });

    it('should prevent adding duplicate allergies', async () => {
      const user = userEvent.setup();
      const mockAddAllergy = jest.fn();

      mockAllergenSearchHook({ allergens: [mockAllergen] });
      renderAllergiesForm({
        selectedAllergies: [mockSelectedAllergy],
        addAllergy: mockAddAllergy,
      });

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(async () => {
        const alreadyAddedOption = screen.getByText(
          'Peanut Allergy (Already added)',
        );
        expect(alreadyAddedOption).toBeInTheDocument();
        await user.click(alreadyAddedOption);
      });

      expect(mockAddAllergy).not.toHaveBeenCalled();
    });
  });

  describe('Input Validation', () => {
    const testCases = [
      { name: 'null selectedItem', selectedItem: null },
      { name: 'undefined selectedItem', selectedItem: undefined },
      {
        name: 'selectedItem with empty uuid',
        selectedItem: {
          uuid: '',
          display: 'Test Allergy',
          type: 'food',
          disabled: false,
        },
      },
      {
        name: 'selectedItem with empty display',
        selectedItem: {
          uuid: 'test-uuid',
          display: '',
          type: 'food',
          disabled: false,
        },
      },
      {
        name: 'selectedItem with missing uuid',
        selectedItem: {
          display: 'Test Allergy',
          type: 'food',
          disabled: false,
        },
      },
      {
        name: 'selectedItem with missing display',
        selectedItem: { uuid: 'test-uuid', type: 'food', disabled: false },
      },
    ];

    testCases.forEach(({ name, selectedItem }) => {
      it(`should not add allergy when ${name}`, () => {
        const mockAddAllergy = jest.fn();
        const { container } = renderAllergiesForm({
          addAllergy: mockAddAllergy,
        });

        const comboBoxElement = container.querySelector('#allergies-search');
        const changeEvent = new CustomEvent('change', {
          detail: { selectedItem },
        });

        comboBoxElement?.dispatchEvent(changeEvent);

        expect(mockAddAllergy).not.toHaveBeenCalled();
      });
    });
  });

  describe('Search Functionality', () => {
    it('should handle input changes and trigger search', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [mockAllergen] });

      renderAllergiesForm();

      const searchBox = getSearchCombobox();
      await user.type(searchBox, 'peanut');

      // Verify search is triggered
      await waitFor(() => {
        expect(screen.getByDisplayValue('peanut')).toBeInTheDocument();
      });
    });

    it('should return empty array when search term is empty', () => {
      renderAllergiesForm();

      // The component should not show any search results when search term is empty
      expect(screen.queryByText('Peanut Allergy')).not.toBeInTheDocument();
    });

    it('should handle search with special characters', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), '!@#$%');

      await waitFor(() => {
        expect(
          screen.getByText(/No matching allergen recorded/i),
        ).toBeInTheDocument();
      });
    });
  });

  describe('ComboBox ItemToString Function', () => {
    it('should format item display with type category', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [mockAllergen] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(() => {
        expect(screen.getByText('Peanut Allergy [Food]')).toBeInTheDocument();
      });
    });

    it('should format item display without type when already selected', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [mockAllergen] });
      renderAllergiesForm({ selectedAllergies: [mockSelectedAllergy] });

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(() => {
        expect(
          screen.getByText('Peanut Allergy (Already added)'),
        ).toBeInTheDocument();
      });
    });

    it('should handle null item in itemToString', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [{ ...mockAllergen, type: null }] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(() => {
        expect(screen.getByText('Peanut Allergy')).toBeInTheDocument();
      });
    });
  });

  describe('Error Handling', () => {
    it('should display error message when API fails', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ error: new Error('API Error'), allergens: [] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'test');

      await waitFor(() => {
        expect(
          screen.getByText(/unexpected error occurred/i),
        ).toBeInTheDocument();
      });
    });

    it('should handle network errors gracefully', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({
        error: new Error('Network Error'),
        allergens: [],
        isLoading: false,
      });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'network');

      await waitFor(() => {
        expect(
          screen.getByText(/unexpected error occurred/i),
        ).toBeInTheDocument();
      });
    });
  });

  describe('Memoization', () => {
    it('should memoize filtered search results correctly', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [mockAllergen] });

      const { rerender } = renderAllergiesForm();

      await user.type(getSearchCombobox(), 'peanut');

      await waitFor(() => {
        expect(screen.getByText('Peanut Allergy [Food]')).toBeInTheDocument();
      });

      // Rerender with same props should use memoized results
      rerender(<AllergiesForm />);

      expect(screen.getByText('Peanut Allergy [Food]')).toBeInTheDocument();
    });
  });

  describe('State Management', () => {
    it('should call removeAllergy when allergy is removed', () => {
      const mockRemoveAllergy = jest.fn();
      renderAllergiesForm({
        selectedAllergies: [mockSelectedAllergy],
        removeAllergy: mockRemoveAllergy,
      });

      const removeButton = screen.getByRole('button', { name: /close/i });
      removeButton.click();

      expect(mockRemoveAllergy).toHaveBeenCalledWith(mockSelectedAllergy.id);
    });

    it('should handle multiple selected allergies', () => {
      const secondAllergy = {
        ...mockSelectedAllergy,
        id: 'test-allergy-2',
        display: 'Shellfish Allergy',
      };

      renderAllergiesForm({
        selectedAllergies: [mockSelectedAllergy, secondAllergy],
      });

      expect(screen.getByText(/Peanut Allergy/)).toBeInTheDocument();
      expect(screen.getByText(/Shellfish Allergy/)).toBeInTheDocument();
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty allergen list gracefully', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [], isLoading: false, error: null });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'nonexistent');

      await waitFor(() => {
        expect(
          screen.getByText(/No matching allergen recorded/i),
        ).toBeInTheDocument();
      });
    });

    it('should handle malformed allergen data', async () => {
      const user = userEvent.setup();
      const malformedAllergen = {
        uuid: 'test-id',
        display: null, // null display
        type: 'food',
        disabled: false,
      };

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      mockAllergenSearchHook({ allergens: [malformedAllergen as any] });

      renderAllergiesForm();

      await user.type(getSearchCombobox(), 'malformed');

      // Should not crash and handle gracefully
      await waitFor(() => {
        expect(getSearchCombobox()).toBeInTheDocument();
      });
    });

    it('should handle rapid search input changes', async () => {
      const user = userEvent.setup();
      mockAllergenSearchHook({ allergens: [mockAllergen] });

      renderAllergiesForm();

      const searchBox = getSearchCombobox();
      await user.type(searchBox, 'p');
      await user.type(searchBox, 'e');
      await user.type(searchBox, 'a');

      await waitFor(() => {
        expect(screen.getByDisplayValue('pea')).toBeInTheDocument();
      });
    });
  });

  describe('Accessibility', () => {
    it('should have no accessibility violations', async () => {
      const { container } = renderAllergiesForm();
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });

  describe('Snapshots', () => {
    it('should match snapshot with no allergies', () => {
      const { container } = renderAllergiesForm();
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with selected allergies', () => {
      const selectedAllergyWithReactions = {
        ...mockSelectedAllergy,
        selectedReactions: [mockReactions[0]],
        hasBeenValidated: true,
      };

      const { container } = renderAllergiesForm({
        selectedAllergies: [selectedAllergyWithReactions],
      });

      expect(container).toMatchSnapshot();
    });
  });
});
