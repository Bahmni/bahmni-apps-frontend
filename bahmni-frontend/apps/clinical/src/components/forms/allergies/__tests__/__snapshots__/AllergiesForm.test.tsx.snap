// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllergiesForm Snapshots should match snapshot with no allergies 1`] = `
<div>
  <div
    class="cds--tile allergiesFormTile"
  >
    <div
      class="allergiesFormTitle"
    >
      Allergies
    </div>
    <div
      class="cds--list-box__wrapper"
    >
      <div
        class="cds--combo-box cds--autoalign cds--list-box cds--list-box--md"
      >
        <div
          class="cds--list-box__field"
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="downshift-«r4i»-menu"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Search for allergies"
            autocomplete="off"
            class="cds--text-input cds--text-input--empty"
            id="allergies-search"
            placeholder="Search to add new allergy"
            role="combobox"
            tabindex="0"
            title=""
            type="text"
            value=""
          />
          <button
            aria-controls="downshift-«r4i»-menu"
            aria-expanded="false"
            aria-label="Open"
            class="cds--list-box__menu-icon"
            id="downshift-«r4i»-toggle-button"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
              />
            </svg>
          </button>
        </div>
        <ul
          aria-labelledby="downshift-«r4i»-label"
          class="cds--list-box__menu"
          id="downshift-«r4i»-menu"
          role="listbox"
          style="position: fixed; left: 0px; top: 0px; visibility: visible; transform: translate(0px, 0px);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`AllergiesForm Snapshots should match snapshot with selected allergies 1`] = `
<div>
  <div
    class="cds--tile allergiesFormTile"
  >
    <div
      class="allergiesFormTitle"
    >
      Allergies
    </div>
    <div
      class="cds--list-box__wrapper"
    >
      <div
        class="cds--combo-box cds--autoalign cds--list-box cds--list-box--md"
      >
        <div
          class="cds--list-box__field"
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-controls="downshift-«r4l»-menu"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Search for allergies"
            autocomplete="off"
            class="cds--text-input cds--text-input--empty"
            id="allergies-search"
            placeholder="Search to add new allergy"
            role="combobox"
            tabindex="0"
            title=""
            type="text"
            value=""
          />
          <button
            aria-controls="downshift-«r4l»-menu"
            aria-expanded="false"
            aria-label="Open"
            class="cds--list-box__menu-icon"
            id="downshift-«r4l»-toggle-button"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
              />
            </svg>
          </button>
        </div>
        <ul
          aria-labelledby="downshift-«r4l»-label"
          class="cds--list-box__menu"
          id="downshift-«r4l»-menu"
          role="listbox"
          style="position: fixed; left: 0px; top: 0px; visibility: visible; transform: translate(0px, 0px);"
        />
      </div>
    </div>
    <div
      aria-label="Added Allergies"
      class="_box_1jr6a_1 allergiesBox cds--css-grid cds--css-grid--narrow cds--css-grid--full-width"
      data-testid="box-w-title"
      role="region"
    >
      <div
        class="_header_1jr6a_8 cds--sm:col-span-4 cds--md:col-span-8 cds--lg:col-span-16 cds--xlg:col-span-16 cds--css-grid-column"
      >
        Added Allergies
      </div>
      <div
        class="cds--sm:col-span-4 cds--md:col-span-8 cds--lg:col-span-16 cds--xlg:col-span-16 cds--css-grid-column"
      >
        <div
          class="_grid_4oucb_1 selectedAllergyItem cds--subgrid cds--subgrid--narrow"
          data-testid="selected-item"
        >
          <div
            class="_children_4oucb_8 cds--sm:col-span-3 cds--md:col-span-7 cds--lg:col-span-15 cds--xlg:col-span-15 cds--css-grid-column"
          >
            <div
              class="cds--subgrid cds--subgrid--wide"
            >
              <div
                class="cds--sm:col-span-4 cds--md:col-span-5 cds--lg:col-span-8 cds--xlg:col-span-8 cds--css-grid-column"
              >
                Peanut Allergy
                 [
                undefined
                ]
                <a
                  class="cds--link"
                  href="#"
                >
                  Add Note
                </a>
              </div>
              <div
                class="cds--sm:col-span-4 cds--md:col-span-3 cds--lg:col-span-3 cds--xlg:col-span-3 cds--css-grid-column"
              >
                <div
                  class="cds--dropdown__wrapper cds--list-box__wrapper"
                >
                  <label
                    class="cds--label cds--visually-hidden"
                    for="downshift-«r4o»-toggle-button"
                    id="downshift-«r4o»-label"
                  >
                    Select Severity
                  </label>
                  <div
                    class="cds--dropdown cds--autoalign cds--list-box"
                    id="allergy-severity-dropdown-test-allergy-1"
                  >
                    <button
                      aria-activedescendant=""
                      aria-controls="downshift-«r4o»-menu"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-label="Select severity for this allergy"
                      class="cds--list-box__field"
                      id="downshift-«r4o»-toggle-button"
                      role="combobox"
                      tabindex="0"
                      title="Select Severity"
                      type="button"
                    >
                      <span
                        class="cds--list-box__label"
                      >
                        Select Severity
                      </span>
                      <div
                        class="cds--list-box__menu-icon"
                      >
                        <svg
                          aria-label="Open menu"
                          fill="currentColor"
                          focusable="false"
                          height="16"
                          name="chevron--down"
                          preserveAspectRatio="xMidYMid meet"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                          />
                          <title>
                            Open menu
                          </title>
                        </svg>
                      </div>
                    </button>
                    <ul
                      aria-labelledby="downshift-«r4o»-label"
                      class="cds--list-box__menu"
                      id="downshift-«r4o»-menu"
                      role="listbox"
                      style="position: fixed; left: 0px; top: 0px; visibility: visible; transform: translate(0px, 0px);"
                    />
                  </div>
                </div>
              </div>
              <div
                class="cds--sm:col-span-4 cds--md:col-span-4 cds--lg:col-span-4 cds--xlg:col-span-4 cds--css-grid-column"
              >
                <div
                  class="cds--multi-select__wrapper cds--multi-select--filterable__wrapper cds--list-box__wrapper cds--autoalign"
                >
                  <label
                    class="cds--label cds--visually-hidden"
                    for="allergy-reactions-multiselect-test-allergy-1-input"
                    id="allergy-reactions-multiselect-test-allergy-1-label"
                  >
                    Select Reactions
                    <span
                      class="cds--visually-hidden"
                    >
                      Total items selected:  1. To clear selection, press Delete or Backspace.
                    </span>
                  </label>
                  <div
                    class="cds--multi-select cds--combo-box cds--multi-select--filterable cds--multi-select--selected cds--list-box"
                    id="allergy-reactions-multiselect-test-allergy-1"
                  >
                    <div
                      class="cds--list-box__field"
                    >
                      <div
                        class="cds--tag cds--tag--filter cds--tag--high-contrast"
                      >
                        <span
                          class="cds--tag__label"
                          title="1"
                        >
                          1
                        </span>
                        <button
                          aria-label="Clear all selected items"
                          class="cds--tag__close-icon"
                          tabindex="-1"
                          title="Clear all selected items"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            fill="currentColor"
                            focusable="false"
                            height="16"
                            preserveAspectRatio="xMidYMid meet"
                            viewBox="0 0 32 32"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
                            />
                          </svg>
                        </button>
                      </div>
                      <input
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        autocomplete="off"
                        class="cds--text-input cds--text-input--empty"
                        id="allergy-reactions-multiselect-test-allergy-1-input"
                        placeholder="Select Reactions"
                        role="combobox"
                        value=""
                      />
                      <button
                        aria-controls="allergy-reactions-multiselect-test-allergy-1__menu"
                        aria-expanded="false"
                        aria-label="Open"
                        class="cds--list-box__menu-icon"
                        id="allergy-reactions-multiselect-test-allergy-1-toggle-button"
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          focusable="false"
                          height="16"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                          />
                        </svg>
                      </button>
                    </div>
                    <ul
                      aria-labelledby="allergy-reactions-multiselect-test-allergy-1-label"
                      class="cds--list-box__menu"
                      id="allergy-reactions-multiselect-test-allergy-1__menu"
                      role="listbox"
                      style="position: fixed; left: 0px; top: 0px; visibility: visible; transform: translate(0px, 0px);"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="_close_4oucb_15 cds--sm:col-span-1 cds--md:col-span-1 cds--lg:col-span-1 cds--xlg:col-span-1 cds--css-grid-column"
          >
            <span
              class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip"
            >
              <div
                class="cds--tooltip-trigger__wrapper"
              >
                <button
                  aria-label="Close Selected Item"
                  class="cds--btn cds--btn--lg cds--layout--size-lg cds--btn--primary cds--btn--icon-only"
                  data-testid="selected-item-close-button"
                  id="close-selected-item"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 32 32"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M17.4141 16L24 9.4141 22.5859 8 16 14.5859 9.4143 8 8 9.4141 14.5859 16 8 22.5859 9.4143 24 16 17.4141 22.5859 24 24 22.5859 17.4141 16z"
                    />
                  </svg>
                </button>
              </div>
              <span
                aria-hidden="true"
                class="cds--popover"
                id="tooltip-«r4t»"
                role="tooltip"
              >
                <span
                  class="cds--popover-content cds--tooltip-content"
                >
                  Selected Item Close
                </span>
                <span
                  class="cds--popover-caret"
                />
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
